pycbc/inference/sampler/ultranest.py

```diff
@@ -0,0 +1,184 @@
+
+    @property
+    def io(self):
+        return self._io
+
+    @property
+    def niterations(self):
+        return self.result['niter']
+
+    @classmethod
+    def from_config(cls, cp, model, **kwds):
+        """
+        Loads the sampler from the given config file.
+        """
+        skeys = {}
+        opts = {'update_interval_iter_fraction': float,
+                'update_interval_ncall': int,
+                'log_interval': int,
+                'show_status': bool,
+                'dlogz': float,
+                'dKL': float,
+                'frac_remain': float,
+                'Lepsilon': float,
+                'min_ess': int,
+                'max_iters': int,
+                'max_ncalls': int,
+                'max_num_improvement_loops': int,
+                'min_num_live_points': int,
+                'cluster_num_live_points:': int}
+        for opt_name in opts:
+            if cp.has_option('sampler', opt_name):
+                value = cp.get('sampler', opt_name)
+                skeys[opt_name] = opts[opt_name](value)
+        return cls(model, **skeys)
+
+    def checkpoint(self):
+        pass
+
+    def resume_from_checkpoint(self):
+        pass
+
+    def finalize(self):
+        logging.info("Writing samples to files")
+        for fn in [self.checkpoint_file, self.backup_file]:
+            self.write_results(fn)
+
+    @property
+    def model_stats(self):
+        return {}
+
+    @property
+    def samples(self):
+        samples_dict = {p: self.result['samples'][:, i] for p, i in
+                        zip(self.model.variable_params, range(self.ndim))}
+        return samples_dict
+
+    def write_results(self, filename):
+        """Writes samples, model stats, acceptance fraction, and random state
+        to the given file.
+
+        Parameters
+        -----------
+        filename : str
+            The file to write to. The file is opened using the ``io`` class
+            in an an append state.
+        """
+        with self.io(filename, 'a') as fp:
+            # write samples
+            fp.write_samples(self.samples, self.model.variable_params)
+            # write log evidence
+            fp.write_logevidence(self.logz, self.logz_err)
+
+    def setup_output(self, output_file):
+        """Sets up the sampler's checkpoint and output files.
+
+        The checkpoint file has the same name as the output file, but with
+        ``.checkpoint`` appended to the name. A backup file will also be
+        created.
+
+        If the output file already exists, an ``OSError`` will be raised.
+        This can be overridden by setting ``force`` to ``True``.
+
+        Parameters
+        ----------
+        sampler : sampler instance
+            Sampler
+        output_file : str
+            Name of the output file.
+        force : bool, optional
+            If the output file already exists, overwrite it.
+        """
+        setup_output(self, output_file)
```

Comment:
Since you're just calling `setup_output` from `sampler.base`, just remove this class method and call it directly in the `from_config`.